packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_upload_task_test.dart

```diff
@@ -244,6 +249,52 @@ void main() {
 
         expect(uploadDataTask.result, throwsA(isA<S3Exception>()));
       });
+
+      test(
+          'cancel() should cancel underlying put object request and throw a S3Exception',
+          () async {
+        const testUploadDataOptions = S3UploadDataOptions(
+          accessLevel: StorageAccessLevel.private,
+          getProperties: true,
+        );
+        final putSmithyOperation = MockSmithyOperation<s3.PutObjectOutput>();
+
+        final completer = Completer<void>();
+        when(
+          () => putSmithyOperation.result,
+        ).thenAnswer((_) async {
+          await completer.future;
+          throw const CancellationException();
+        });
+        when(
+          putSmithyOperation.cancel,
+        ).thenAnswer((_) async {});
+        when(() => putSmithyOperation.requestProgress)
+            .thenAnswer((_) => Stream.value(1));
+
+        when(
+          () => s3Client.putObject(any()),
+        ).thenAnswer((_) => putSmithyOperation);
+
+        final uploadDataTask = S3UploadTask.fromDataPayload(
+          testDataPayload,
+          s3Client: s3Client,
+          prefixResolver: testPrefixResolver,
+          bucket: testBucket,
+          key: testKey,
+          options: testUploadDataOptions,
+          logger: logger,
+          transferDatabase: transferDatabase,
+        );
+
+        unawaited(uploadDataTask.start());
+        await uploadDataTask.cancel();
+
+        completer.complete();
+
+        await expectLater(uploadDataTask.result, throwsA(isA<S3Exception>()));
```

Comment:
Can we test that this is throwing a CancellationException?

Reply:
`await UploadTask.result` will never throw `CancellationException`.  `CancellationException` may be only caught internally on pause/cancel.

On pause, `UploadTask.result` won't complete and waiting for resume.
On cancel, `UploadTask.result` completes with exception `S3Exception.controllableOperationCanceled()` which is verified by this test.